Add new config (non-dynamic) for  agent connections monitor thread, and keep timeunit to secs (in sync with the earlier Wait config)

sureshanaparti opened the pull request:
### Description

This PR adds new config 'agent.connections.monitor.interval' (non-dynamic) to monitor and cleanup new agent connections, instead of using 'Wait' which is dynamic, and updates the monitor thread timeunit to secs, in sync with the earlier Wait config.

New agent connections monitor thread introduced here: https://github.com/apache/cloudstack/pull/9840


<!--- Describe your changes in DETAIL - And how has behaviour functionally changed. -->

<!-- For new features, provide link to FS, dev ML discussion etc. -->
<!-- In case of bug fix, the expected and actual behaviours, steps to reproduce. -->

<!-- When "Fixes: #<id>" is specified, the issue/PR will automatically be closed when this PR gets merged -->
<!-- For addressing multiple issues/PRs, use multiple "Fixes: #<id>" -->
<!-- Fixes: # -->

<!--- ******************************************************************************* -->
<!--- NOTE: AUTOMATION USES THE DESCRIPTIONS TO SET LABELS AND PRODUCE DOCUMENTATION. -->
<!--- PLEASE PUT AN 'X' in only **ONE** box -->
<!--- ******************************************************************************* -->

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### Screenshots (if appropriate):

### How Has This Been Tested?

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->

#### How did you try to break this feature and the system with this change?

<!-- see how your change affects other areas of the code, etc. -->

<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/main/CONTRIBUTING.md) document -->

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 12699

sureshanaparti commented:
@blueorangutan test

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-12616)<b/>

sureshanaparti commented:
@blueorangutan test

blueorangutan commented:
@sureshanaparti a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

DaanHoogland reviewed (approved):
clgtm

blueorangutan commented:
<b>[SF] Trillian test result (tid-12633)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 56917 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr10525-t12633-kvm-ol8.zip
Smoke tests completed. 140 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
ContextSuite context=TestSharedNetworkWithConfigDrive>:setup | `Error` | 1521.08 | test_network.py

Pearl1594 reviewed (approved):
CLGTM

shwstppr reviewed (commented):
@sureshanaparti sorry it was my mistake considering `wait` config in minutes but can we also add the following change to fix cleanup time in seconds in the task and some optimisation. I'm happy to open a separate PR if needed

```
diff --git a/engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java b/engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java
index 818a3c5a23..1504af702b 100644
--- a/engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java
+++ b/engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java
@@ -51,8 +51,8 @@ import org.apache.cloudstack.framework.jobs.AsyncJobExecutionContext;
 import org.apache.cloudstack.managed.context.ManagedContextRunnable;
 import org.apache.cloudstack.outofbandmanagement.dao.OutOfBandManagementDao;
 import org.apache.cloudstack.utils.identity.ManagementServerNode;
-import org.apache.commons.collections.MapUtils;
 import org.apache.cloudstack.utils.reflectiontostringbuilderutils.ReflectionToStringBuilderUtils;
+import org.apache.commons.collections.MapUtils;
 import org.apache.commons.lang3.BooleanUtils;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.logging.log4j.ThreadContext;
@@ -1859,25 +1859,19 @@ public class AgentManagerImpl extends ManagerBase implements AgentManager, Handl
             logger.trace("Agent New Connections Monitor is started.");
             final int cleanupTime = Wait.value();
             Set<Map.Entry<String, Long>> entrySet = newAgentConnections.entrySet();
-            long cutOff = System.currentTimeMillis() - (cleanupTime * 60 * 1000L);
-            if (logger.isDebugEnabled()) {
-                List<String> expiredConnections = newAgentConnections.entrySet()
-                        .stream()
-                        .filter(e -> e.getValue() <= cutOff)
-                        .map(Map.Entry::getKey)
-                        .collect(Collectors.toList());
-                logger.debug(String.format("Currently %d active new connections, of which %d have expired - %s",
-                        entrySet.size(),
-                        expiredConnections.size(),
-                        StringUtils.join(expiredConnections)));
-            }
-            for (Map.Entry<String, Long> entry : entrySet) {
-                if (entry.getValue() <= cutOff) {
-                    if (logger.isTraceEnabled()) {
-                        logger.trace(String.format("Cleaning up new agent connection for %s", entry.getKey()));
-                    }
-                    newAgentConnections.remove(entry.getKey());
-                }
+            long cutOff = System.currentTimeMillis() - (cleanupTime * 1000L);
+            List<String> expiredConnections = newAgentConnections.entrySet()
+                    .stream()
+                    .filter(e -> e.getValue() <= cutOff)
+                    .map(Map.Entry::getKey)
+                    .collect(Collectors.toList());
+            logger.debug("Currently {} active new connections, of which {} have expired - {}",
+                    entrySet.size(),
+                    expiredConnections.size(),
+                    StringUtils.join(expiredConnections));
+            for (String connection : expiredConnections) {
+                logger.trace("Cleaning up new agent connection for {}", connection);
+                newAgentConnections.remove(connection);
             }
         }
     }
```

sureshanaparti commented:
> @sureshanaparti sorry it was my mistake considering `wait` config in minutes but can we also add the following change to fix cleanup time in seconds in the task and some optimisation. I'm happy to open a separate PR if needed
> 

updated @shwstppr, missed this, thanks for pointing out.

sureshanaparti commented:
@blueorangutan package

blueorangutan commented:
@sureshanaparti a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 12804

Pearl1594 commented:
@blueorangutan test

blueorangutan commented:
@Pearl1594 a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-12731)<b/>

sureshanaparti commented:
@blueorangutan test

blueorangutan commented:
@sureshanaparti a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-12785)<b/>

sureshanaparti commented:
@blueorangutan package

blueorangutan commented:
@sureshanaparti a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

sureshanaparti commented:
@blueorangutan package

blueorangutan commented:
@sureshanaparti a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 13035

sureshanaparti commented:
@blueorangutan test

blueorangutan commented:
@sureshanaparti a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-12976)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 53944 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr10525-t12976-kvm-ol8.zip
Smoke tests completed. 141 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

DaanHoogland commented:
@sureshanaparti @Pearl1594 , I added needs-testing to this PR, but I am not really convinced if we need to and how. Can you advise?

sureshanaparti commented:
> @sureshanaparti @Pearl1594 , I added needs-testing to this PR, but I am not really convinced if we need to and how. Can you advise?

@DaanHoogland Update Wait config, to say 120 secs, enable trace log and check for the log "_Agent New Connections Monitor is started_", it has to be logged every 120 secs.

sureshanaparti commented:
> > @sureshanaparti @Pearl1594 , I added needs-testing to this PR, but I am not really convinced if we need to and how. Can you advise?
> 
> @DaanHoogland Update Wait config, to say 120 secs, enable trace log and check for the log "_Agent New Connections Monitor is started_", it has to be logged every 120 secs.

@DaanHoogland The agent connections monitor task required restart, I've updated with new config (non-dynamic) to monitor and cleanup new agent connections, instead of using Wait which is dynamic.

blueorangutan commented:
@sureshanaparti a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

DaanHoogland commented:
tested:
set ‘agent.connections.monitor.interval’ to 59 seconds and had to restart to see it effected,
set ‘wait’ to 61 seconds, no restart adviced,
'Agent New Connections Monitor is started’ messages apeared aproximately every 59 seconds.
